libtenzir/include/tenzir/format/syslog.hpp

```diff
@@ -250,6 +254,134 @@ struct message_parser : parser_base<message_parser> {
   }
 };
 
+/// A legacy (RFC 3164) Syslog message.
+struct legacy_message {
+  std::optional<uint16_t> facility;
+  std::optional<uint16_t> severity;
+  std::string timestamp;
+  std::string host;
+  std::string tag;
+  std::string content;
+};
+
+/// Parser for legacy Syslog messages.
+/// @relates legacy_message
+struct legacy_message_parser : parser_base<legacy_message_parser> {
+  using attribute = legacy_message;
+
+  template <typename Iterator, typename Attribute>
+  bool parse(Iterator& f, const Iterator& l, Attribute& x) const {
+    using namespace parsers;
+    using namespace parser_literals;
+    const auto is_prival = [](uint16_t in) {
+      return in <= 191;
+    };
+    const auto to_facility_and_severity = [&](uint16_t in) {
+      if constexpr (!std::is_same_v<Attribute, unused_type>) {
+        x.facility = in / 8;
+        x.severity = in % 8;
+      }
+    };
+    const auto prival = ignore(
+      integral_parser<uint16_t, 3>{}.with(is_prival)->*to_facility_and_severity);
+    const auto priority_parser = '<' >> prival >> '>';
+    const auto word = +(parsers::printable - space);
+    const auto ws = +space;
+    const auto wsignore = ignore(ws);
+    const auto is_month = [](const std::string& mon) {
+      return mon == "Jan" || mon == "Feb" || mon == "Mar" || mon == "Apr"
+             || mon == "May" || mon == "Jun" || mon == "Jul" || mon == "Aug"
+             || mon == "Sep" || mon == "Oct" || mon == "Nov" || mon == "Dec";
```

Comment:
Do we want such strong validation (not only for months, but in general)? This is not a rhetoric question, I'm not sure what's best here. I guess we would need some kind of heuristic if for example the date is not in a format that we expect (e.g., everything after priority and up to the second-to-last word before the colon). And that might also break if we want to transform this to an actual timestamp using the current year.

Reply:
It's a fine line. We need to do _some_ validation, because the lenient parser we had initially

```cpp
auto timestamp = word >> ws >> word >> ws >> word;
auto p = ~(priority >> wsignore)
         >> timestamp >> wsignore
         >> word >> wsignore        // host
         >> word >> ':'             // tag
         >> *parsers::printable;
```

would accept inputs that are clearly not conformant to the old-rfc format, which was caught by our integration tests:

```
# Malformed new-rfc syslog message
2003-10-11T22:14:15.003Z mymachineexamplecom su asd ID47: - Message

# Would nevertheless be parsed as the following old-rfc message:
{
  "facility": null,
  "severity": null,
  "timestamp": "2003-10-11T22:14:15.003Z mymachineexamplecom su",
  "host": "asd",
  "tag": "ID47",
  "content": "- Message"
}
```

I think it's better that those messages are filed under the `syslog.unknown` schema, with everything being in one `"syslog_message"` field, rather than the result being poorly structured like above.

Reply:
That's a great argument :+1: